pkg/controller/user-cluster-controller-manager/flatcar/flatcar_controller.go

```diff
@@ -64,12 +63,7 @@ func Add(mgr manager.Manager, overwriteRegistry string, updateWindow kubermaticv
 		return err
 	}
 
-	predicate := predicateutil.Factory(func(o ctrlruntimeclient.Object) bool {
-		node := o.(*corev1.Node)
-		return o.GetLabels()[nodelabelerapi.DistributionLabelKey] == nodelabelerapi.FlatcarLabelValue && !node.Spec.Unschedulable
-	})
-
-	return c.Watch(&source.Kind{Type: &corev1.Node{}}, controllerutil.EnqueueConst(""), predicate)
```

Comment:
Umm, I am not a big fan of removing the predicate all together. Since that would result in a lot of redundant reconciliations.

For reate and delete events it's fine because they don't occur much often. But update is something that can happen quite frequently. I think for updates, we can still keep this predicate or at the very least keep `o.GetLabels()[nodelabelerapi.DistributionLabelKey] == nodelabelerapi.FlatcarLabelValue`

Reply:
What happens when the `user-cluster-controller-manager` is down while the last Flatcar node is removed? Then the removal logic will be never be triggered as when it starts back up, the predicate filters out all of the existing nodes.

Given that we only do `LIST` and `GET` within the reconcile loop and that will work with the cache afaik, the redundant reconciliations are not a problem IMHO.

Reply:
When `user-cluster-controller-manager` will start again, it will re-queue existing nodes and that should be sufficient :) 

> Given that we only do LIST and GET within the reconcile loop and that will work with the cache afaik, the redundant reconciliations are not a problem IMHO.

For every update on nodes whether it's flatcar or not, you'll be triggering a reconcile loop. Doesn't sound logically correct to me at least.

Reply:
> When user-cluster-controller-manager will start again, it will re-queue existing nodes and that should be sufficient :)

What I mean is the scenario that this PR is trying to address: The last Flatcar node is removed from the cluster while `user-cluster-controller-manager` is down. Since the node is gone after being removed, `user-cluster-controller-manager` will not be aware of it, and the predicate will keep the removal logic from being triggered.

Reply:
As per our chat, I kept a predicate in but limited the label check for `UPDATE` events. `CREATE` and `DELETE` events are triggering the reconcile checks unconditionally. Works with starting up `user-cluster-controller-manager` after removing all Flatcar nodes.